### PR TITLE
Add «resize to fill box» alternative for image endpoints

### PR DESCRIPTION
**Pull request by oddstr13**
**Changes**

Adds fillHeight and fillWidth optional parameters to all(?) image endpoints.

**Issues**

Server portion for resolving jellyfin/jellyfin-web#2191

Web PR: jellyfin/jellyfin-web#2514
apiClient PR: jellyfin/jellyfin-apiclient-javascript#167

### CODE REVIEW

File: MediaBrowser.Model/Drawing/DrawingUtils.cs [hunk 1]
```diff
@@ -57,6 +57,54 @@ public static class DrawingUtils
...
+        {
+            // Return original size if input is invalid.
+            if (
+                (fillWidth == null && fillHeight == null)
+                || (fillWidth == 0 || fillHeight == 0))
+            {
+                return size;
+            }
+
+            if (fillWidth == null || fillWidth == 0)
```
Comment (cvium) [inline MediaBrowser.Model/Drawing/DrawingUtils.cs:81]: This is never true, is it?

File: MediaBrowser.Model/Drawing/DrawingUtils.cs [hunk 2]
```diff
@@ -57,6 +57,54 @@ public static class DrawingUtils
...
+            }
+
+            if (fillHeight == null || fillHeight == 0)
+            {
+                fillHeight = 1;
+            }
+
+            double widthRatio = (double)size.Width / (double)fillWidth;
+            double heightRatio = (double)size.Height / (double)fillHeight!;
+            // min()
```
Comment (cvium) [inline MediaBrowser.Model/Drawing/DrawingUtils.cs:93]: We do have `Math.Min` that has the exact same implementation

File: MediaBrowser.Model/Drawing/DrawingUtils.cs (hunk 1 above)
Comment (oddstr13) [reply MediaBrowser.Model/Drawing/DrawingUtils.cs:81]: I see that my logic in the block above is not entirely correct, but this one is true whenever height is specified but not width.

File: MediaBrowser.Model/Drawing/DrawingUtils.cs (hunk 2 above)
Comment (oddstr13) [reply MediaBrowser.Model/Drawing/DrawingUtils.cs:93]: I didn't find the proper documentation for Math.Min last night, and what I found implied it was int only, thanks for making me take another look :P

**Comment by Bond-009** [comment]
Why backport this?

**Comment by oddstr13** [comment]
Because it is a major UX improvement (once the corresponding changes in web are done) for something that gets brought up relatively often.

image
This is what cards with resized images currently look like when the aspect ratio doesn't match.
Larger mismatch equals worse image quality.

I also think the code paths here have a relatively low risk of breakage, but please do test my changes well before merging (looks fine so far in my own testing).

**Comment by oddstr13** [comment]
This should be seen as the server part of fixing a UX regression introduced by jellyfin/jellyfin-web#1710

File: MediaBrowser.Model/Drawing/DrawingUtils.cs [hunk 3]
```diff
@@ -57,6 +57,52 @@ public static class DrawingUtils
...
+                fillWidth = 1;
+            }
+
+            if (fillHeight == null || fillHeight == 0)
+            {
+                fillHeight = 1;
+            }
+
+            double widthRatio = (double)size.Width / (double)fillWidth;
+            double heightRatio = (double)size.Height / (double)fillHeight;
```
Comment (crobibero) [inline MediaBrowser.Model/Drawing/DrawingUtils.cs:91]:
```suggestion
            double widthRatio = size.Width / (double)fillWidth;
            double heightRatio = size.Height / (double)fillHeight;
```

File: MediaBrowser.Model/Drawing/DrawingUtils.cs [hunk 4]
```diff
@@ -57,6 +57,52 @@ public static class DrawingUtils
...
+            double scaleRatio = Math.Min(widthRatio, heightRatio);
+
+            // Clamp to current size.
+            if (scaleRatio < 1)
+            {
+                return size;
+            }
+
+            int newWidth = Convert.ToInt32(Math.Ceiling((double)size.Width / scaleRatio));
+            int newHeight = Convert.ToInt32(Math.Ceiling((double)size.Height / scaleRatio));
```
Comment (crobibero) [inline MediaBrowser.Model/Drawing/DrawingUtils.cs:101]:
```suggestion
            int newWidth = Convert.ToInt32(Math.Ceiling(size.Width / scaleRatio));
            int newHeight = Convert.ToInt32(Math.Ceiling(size.Height / scaleRatio));
```

**Review by crobibero — APPROVED** [review]
Some casts aren't needed

File: MediaBrowser.Model/Drawing/DrawingUtils.cs (hunk 4 above)
Comment (oddstr13) [reply MediaBrowser.Model/Drawing/DrawingUtils.cs:101]: So, division with int is implicitly done using doubles? What would `(int)123 / (int)100` return?

File: Emby.Drawing/ImageProcessor.cs [hunk 5]
```diff
@@ -246,48 +261,111 @@ private ImageFormat GetOutputFormat(IReadOnlyCollection<ImageFormat> clientSuppo
...
+            string backgroundColor,
+            string foregroundLayer)
         {
-            var filename = originalPath
-                + "width=" + outputSize.Width
-                + "height=" + outputSize.Height
-                + "quality=" + quality
-                + "datemodified=" + dateModified.Ticks
-                + "f=" + format;
+            System.Text.StringBuilder filename = new System.Text.StringBuilder(128);
```
Comment (cvium) [inline Emby.Drawing/ImageProcessor.cs:282]:
```suggestion
            var filename = new StringBuilder(128);
```

Comment (cvium) [reply Emby.Drawing/ImageProcessor.cs:282]: Have to import `System.Text` though.

File: MediaBrowser.Controller/Drawing/ImageProcessingOptions.cs [hunk 6]
```diff
@@ -95,6 +99,16 @@ public bool HasDefaultOptions(string originalImagePath, ImageDimensions? size)
                 return false;
             }
 
+            if (FillWidth.HasValue && sizeValue.Width > FillWidth.Value)
+            {
+                return false;
+            }
+
+            if (FillHeight.HasValue && sizeValue.Height > FillHeight.Value)
```
Comment (cvium) [inline MediaBrowser.Controller/Drawing/ImageProcessingOptions.cs:107]: Comparisons with null return false.
```suggestion
            if (sizeValue.Width > FillWidth || sizeValue.Height > FillHeight)
```

File: Emby.Drawing/ImageProcessor.cs (hunk 5 above)
Comment (oddstr13) [reply Emby.Drawing/ImageProcessor.cs:282]: I wasn't quite sure I wanted to import `System.Text` just for `StringBuilder`. Should I always do that?

Comment (cvium) [reply Emby.Drawing/ImageProcessor.cs:282]: There's no functional difference

Comment (oddstr13) [reply Emby.Drawing/ImageProcessor.cs:282]: Well, the functional difference would be bringing all of `System.Text.*` into scope, no?

Comment (cvium) [reply Emby.Drawing/ImageProcessor.cs]: I suppose, but the IL remains the same

**Review by cvium — APPROVED** [review]

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**